tests/plugins/test_koji_upload.py

```diff
@@ -0,0 +1,1092 @@
+        'kojihub': '',
+        'url': '/',
+    }
+    if ssl_certs:
+        args['koji_ssl_certs'] = '/'
+
+    if principal:
+        args['koji_principal'] = principal
+
+    if keytab:
+        args['koji_keytab'] = keytab
+
+    if metadata_only:
+        args['metadata_only'] = True
+
+    if blocksize:
+        args['blocksize'] = blocksize
+
+    if target:
+        args['target'] = target
+        args['poll_interval'] = 0
+
+    plugins_conf = [
+        {'name': KojiUploadPlugin.key, 'args': args},
+    ]
+
+    if target and tag_later:
+        plugins_conf.append({'name': KojiTagBuildPlugin.key,
+                             'args': {
+                                 'kojihub': '',
+                                 'target': target,
+                                 'poll_interval': 0.01}})
+    workflow.postbuild_plugins_conf = plugins_conf
+    runner = PostBuildPluginsRunner(tasker, workflow, plugins_conf)
+    return runner
+
+def get_metadata(workflow, osbs):
+    cm_annotations = workflow.postbuild_results[KojiUploadPlugin.key]
+
+    assert "metadata_fragment" in cm_annotations
+    assert "metadata_fragment_key" in cm_annotations
+
+    cmlen = len("configmap/")
+    cm_key = cm_annotations["metadata_fragment"][cmlen:]
+
+    cm_frag_key = cm_annotations["metadata_fragment_key"]
+    cm_data = osbs.get_config_map(cm_key)
+
+    return cm_data[cm_frag_key]
+
+class TestKojiUploadLogger(object):
+    @pytest.mark.parametrize('totalsize', [0, 1024])
+    def test_with_zero(self, totalsize):
+        logger = flexmock()
+        logger.should_receive('debug').once()
+        upload_logger = KojiUploadLogger(logger)
+        upload_logger.callback(0, totalsize, 0, 0, 0)
+
+    @pytest.mark.parametrize(('totalsize', 'step', 'expected_times'), [
+        (10, 1, 11),
+        (12, 1, 7),
+        (12, 3, 5),
+    ])
+    def test_with_defaults(self, totalsize, step, expected_times):
+        logger = flexmock()
+        logger.should_receive('debug').times(expected_times)
+        upload_logger = KojiUploadLogger(logger)
+        upload_logger.callback(0, totalsize, 0, 0, 0)
+        for offset in range(step, totalsize + step, step):
+            upload_logger.callback(offset, totalsize, step, 1.0, 1.0)
+
+    @pytest.mark.parametrize(('totalsize', 'step', 'notable', 'expected_times'), [
+        (10, 1, 10, 11),
+        (10, 1, 20, 6),
+        (10, 1, 25, 5),
+        (12, 3, 25, 5),
+    ])
+    def test_with_notable(self, totalsize, step, notable, expected_times):
+        logger = flexmock()
+        logger.should_receive('debug').times(expected_times)
+        upload_logger = KojiUploadLogger(logger, notable_percent=notable)
+        for offset in range(0, totalsize + step, step):
+            upload_logger.callback(offset, totalsize, step, 1.0, 1.0)
+
+
+class TestKojiUpload(object):
+    def test_koji_upload_failed_build(self, tmpdir, os_env):
+        session = MockedClientSession('')
+        osbs = MockedOSBS()
```

Comment:
Hi Vadim,

That should have been resolved in the latest version late last week.